lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -1302,6 +1304,278 @@ void ESItoHWPass::runOnOperation() {
+    e.value().receiving().replaceAllUsesWith(
+        implOp.getResult(e.index() + instOpNumResults));
+
+  instOp.erase();
+  return success();
+}
+
+LogicalResult ESIConnectServicesPass::surfaceReqs(
+    hw::HWModuleLike mod, ArrayRef<RequestToClientConnection> toClientReqs,
+    ArrayRef<RequestToServerConnection> toServerReqs) {
+  auto ctxt = mod.getContext();
+  Block &modBlock = mod->getRegion(0).front();
+  Operation *modTerminator = modBlock.getTerminator();
+
+  // Track initial operand/result counts and the new IO.
+  unsigned origNumInputs = getModuleNumInputs(mod);
+  SmallVector<std::pair<unsigned, hw::PortInfo>> newInputs;
+  unsigned origNumOutputs = getModuleNumOutputs(mod);
+  SmallVector<std::pair<unsigned, hw::PortInfo>> newOutputs;
+
+  // Assemble a port name from an array.
+  auto getPortName = [](ArrayAttr namePath) {
+    std::string portName;
+    llvm::raw_string_ostream nameOS(portName);
+    llvm::interleave(
+        namePath.getValue(), nameOS,
+        [&](Attribute attr) { nameOS << attr.cast<StringAttr>().getValue(); },
+        ".");
+    return nameOS.str();
+  };
+
+  // Append input ports to new port list and replace uses with new block args
+  // which will correspond to said ports.
+  unsigned inputCounter = origNumInputs;
+  for (auto toClient : toClientReqs) {
+    newInputs.push_back(std::make_pair(
+        origNumInputs,
+        hw::PortInfo{
+            StringAttr::get(ctxt, getPortName(toClient.clientNamePath())),
+            hw::PortDirection::INPUT, toClient.getType(), inputCounter++}));
+    toClient.replaceAllUsesWith(modBlock.addArgument(
+        toClient.receiving().getType(), toClient.getLoc()));
+  }
+  // Append output ports to new port list and redirect toServer inputs to output
+  // op.
+  unsigned outputCounter = origNumOutputs;
+  for (auto toServer : toServerReqs) {
+    newOutputs.push_back(std::make_pair(
+        origNumOutputs,
+        hw::PortInfo{
+            StringAttr::get(ctxt, getPortName(toServer.clientNamePath())),
+            hw::PortDirection::OUTPUT, toServer.sending().getType(),
+            outputCounter++}));
+    modTerminator->insertOperands(modTerminator->getNumOperands(),
+                                  toServer.sending());
+  }
+
+  // Modify the module ports.
+  hw::modifyModulePorts(mod, newInputs, newOutputs, {}, {});
+
+  // Prepend a name to the instance tracking array.
+  auto prependNamePart = [&](ArrayAttr namePath, StringAttr part) {
+    SmallVector<Attribute, 8> newNamePath;
+    newNamePath.push_back(part);
+    newNamePath.append(namePath.begin(), namePath.end());
+    return ArrayAttr::get(namePath.getContext(), newNamePath);
+  };
+
+  // Update the module instantiations.
+  SmallVector<hw::HWInstanceLike, 1> newModuleInstantiations;
+  for (auto inst : moduleInstantiations[mod]) {
+    OpBuilder b(inst);
+
+    auto hwInst = dyn_cast<hw::InstanceOp>(inst.getOperation());
+    if (!hwInst)
+      return inst.emitError(
```

Comment:
nit: missing test

Reply:
This error won't be there for long, so I'm not testing it.